README.md
# Rotating Icon Button | Flutter Package

[Flutter](https://flutter.dev/) package for rotating and smoothly animating any child widget of its own when tapped. The package has been written solely in [Dart](https://dart.dev/) language.

&nbsp;

| ![rotating_icon_button_gif_1](https://github.com/BBarisKilic/Rotating-Icon-Button/blob/master/screenshot/rotating_icon_button_gif_1.gif?raw=true) | ![rotating_icon_button_gif_4](https://github.com/BBarisKilic/Rotating-Icon-Button/blob/master/screenshot/rotating_icon_button_gif_4.gif?raw=true) |  ![rotating_icon_button_gif_3](https://github.com/BBarisKilic/Rotating-Icon-Button/blob/master/screenshot/rotating_icon_button_gif_3.gif?raw=true)
| :-: | :-: | :-: |

&nbsp;

## Features

A powerful and fully customizable widget. With this package, you can easily create live buttons that rotate in a direction you want, in a duration you assign, etc.

`RotatingIconButton` widget allows you to create living buttons to make your UI look lovely.

&nbsp;

<p align="center"><img src="https://github.com/BBarisKilic/Rotating-Icon-Button/blob/master/screenshot/rotating_icon_button_gif_2.gif?raw=true" width="300"></p>

&nbsp;

## Getting started

Let's take a look at how to implement `RotatingIconButton` to create lovely animating buttons.

First, add the following line to `pubspec.yaml`:
```yaml
rotating_icon_button: ^1.0.0
```

Second, import `RotatingIconButton`:
```dart
import 'package:rotating_icon_button/rotating_icon_button.dart';
```

## Usage

Basic usage example: 

`main.dart`
```dart
// Import RotatingIconButton
import 'package:rotating_icon_button/rotating_icon_button.dart';

@override
Widget build(BuildContext context) {
    // Call the RotatingIconButton widget anywhere in your widget tree.
    return RotatingIconButton(
      onTap: () {},
      elevation: 10.0,
      shadowColor: Colors.pink,
      borderRadius: 20.0,
      rotateType: RotateType.semi,
      clockwise: false,
      padding: const EdgeInsets.symmetric(
        vertical: 10,
        horizontal: 40,
      ),
      background: Colors.grey,
      child: const Icon(
        Icons.navigation,
        color: Colors.pink,
      ),
    );
}
```

&nbsp;

## Example

More examples can be found in `/example` folder on [GitHub](https://github.com/BBarisKilic/Rotating-Icon-Button). 

[Media Player](https://github.com/BBarisKilic/Rotating-Icon-Button/tree/master/example/media_player) - an example of basic implementation of `RotatingIconButton`. Simple project that simulates a media player. As usual, the player needs backward and forward buttons. Thanks to the `RotatingIconButton` package, it is so easy to create buttons with rotating animations that perfectly fit the situation.

[Mixed Buttons](https://github.com/BBarisKilic/Rotating-Icon-Button/tree/master/example/mixed_buttons) - an example of basic implementation of `RotatingIconButton`. Simple project that shows many `RotatingIconButton` in different types and shapes. It is mainly created to show you the power of this widget and help you to understand how to customize your buttons through using this widget.

&nbsp;

## Additional information

This package has been written solely in Dart language, and it is not using any third-party package as a dependency.

For more information please visit [GitHub](https://github.com/BBarisKilic/Rotating-Icon-Button).

&nbsp;

## Feature requests and bugs

Please file feature requests and bugs at the [issue tracker](https://github.com/BBarisKilic/Rotating-Icon-Button/issues).

&nbsp;

## Maintainers

- [Bulent Baris Kilic](https://github.com/BBarisKilic)

CHANGELOG.md
## 1.0.0

* Initial release of `rotating_icon_button`

test/rotating_button_test.dart
// Copyright 2022 The RotatingIconButton Author. All rights reserved.
// Use of this source code is governed by a MIT-style license that
// can be found in the LICENSE file.

import 'package:flutter_test/flutter_test.dart';

void main() {
  test('test', () {});
}


example/media_player/test/widget_test.dart
// Copyright 2022 The RotatingIconButton Author. All rights reserved.
// Use of this source code is governed by a MIT-style license that
// can be found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:media_player/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (tester) async {
    await tester.pumpWidget(const ExampleApp());

    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


example/mixed_buttons/test/widget_test.dart
// Copyright 2022 The RotatingIconButton Author. All rights reserved.
// Use of this source code is governed by a MIT-style license that
// can be found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:mixed_buttons/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (tester) async {
    await tester.pumpWidget(const ExampleApp());

    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/rotating_button.dart
// Copyright 2022 The RotatingButton Author. All rights reserved.
// Use of this source code is governed by a MIT-style license that
// can be found in the LICENSE file.

library rotating_button;

import 'dart:math';

import 'package:flutter/material.dart';

class RotatingButton extends StatefulWidget {
  const RotatingButton({
    Key? key,
    required this.onTap,
    required this.child,
    this.duration = const Duration(milliseconds: 300),
    this.curve = Curves.easeInOut,
    this.degrees = 90.0,
    this.clockwise = true,
  }) : super(key: key);

  final void Function() onTap;
  final Widget child;
  final Duration duration;
  final Curve curve;
  final double degrees;
  final bool clockwise;

  @override
  State<RotatingButton> createState() => _RotatingButtonState();
}

class _RotatingButtonState extends State<RotatingButton>
    with SingleTickerProviderStateMixin {
  late final AnimationController controller;
  late final Animation<double> animation;

  @override
  void initState() {
    super.initState();
    initController();
    addListener();
    initAnimation(degrees: widget.degrees, clockwise: widget.clockwise);
  }

  void initController() {
    controller = AnimationController(
      vsync: this,
      duration: widget.duration,
      reverseDuration: widget.duration,
    );
  }

  void addListener() {
    controller.addStatusListener(listener);
  }

  void listener(AnimationStatus status) {
    if (mounted) {
      if (status == AnimationStatus.completed) {
        controller.reverse();
      }
    }
  }

  @override
  void dispose() {
    controller
      ..removeStatusListener(listener)
      ..dispose();
    super.dispose();
  }

  void initAnimation({required double degrees, required bool clockwise}) {
    final double angle = degrees * pi / 180;

    animation =
        Tween<double>(begin: 0, end: clockwise ? angle : -angle).animate(
      CurvedAnimation(
        parent: controller,
        curve: widget.curve,
      ),
    );
  }

  void onTap() {
    if (mounted) {
      widget.onTap();
      controller.forward(from: 0.0);
    }
  }

  @override
  Widget build(BuildContext context) => GestureDetector(
        onTap: onTap,
        child: AnimatedBuilder(
          animation: animation,
          builder: (context, child) =>
              Transform.rotate(angle: animation.value, child: widget.child),
        ),
      );
}


lib/rotating_icon_button.dart
// Copyright 2022 The RotatingIconButton Author. All rights reserved.
// Use of this source code is governed by a MIT-style license that
// can be found in the LICENSE file.

library rotating_icon_button;

import 'dart:math';

import 'package:flutter/material.dart';

/// An enum to make it easier to set the rotation type of the child widget.
///
/// There are three different rotation types:
///
/// 1- [RotateType.full]
///
/// 2- [RotateType.semi]
///
/// 3- [RotateType.quarter]
enum RotateType {
  /// [RotateType.quarter] => To rotate the child widget 90 degrees
  quarter,

  /// [RotateType.semi] => To rotate the child widget 180 degrees
  semi,

  /// [RotateType.full] => To rotate the child widget completely.
  full,
}

/// An enum to make it easier to set the shape of the button.
///
/// There are two different shapes:
///
/// 1- [ButtonShape.rectangle]
///
/// 2- [ButtonShape.circle]
enum ButtonShape {
  /// [ButtonShape.rectangle] => To give this button a rectangle or
  /// square shape.
  rectangle,

  /// [ButtonShape.circle] => To give this button a circle shape.
  circle,
}

/// [RotatingIconButton] Widget
///
/// Flutter package for rotating and smoothly animating any child
/// widget of its own when tapped.
///
/// The package has been written solely in Dart language.
class RotatingIconButton extends StatefulWidget {
  /// Create [RotatingIconButton] widget.
  ///
  /// A widget for rotating and smoothly animating any child widget
  /// of its own when tapped.
  const RotatingIconButton({
    Key? key,
    required this.onTap,
    required this.child,
    this.padding,
    this.rotateType = RotateType.quarter,
    this.clockwise = true,
    this.duration = const Duration(milliseconds: 300),
    this.curve = Curves.easeInOut,
    this.shape = ButtonShape.rectangle,
    this.elevation = 0.0,
    this.shadowColor,
    this.background,
    this.borderRadius,
  }) : super(key: key);

  /// [onTap] Function that provides you the ability to control tap action.
  ///
  /// With this function, you can easily assign tasks to the widget.
  final void Function() onTap;

  /// This widget will be displayed inside the button.
  ///
  /// It is required to provide a widget to show it as part of the button.
  /// Normally Icon widget is expected as a [child] widget but it is not
  /// limited only to it.
  final Widget child;

  /// Add [padding] to the [child] widget to customize the button.
  ///
  /// The default value is "const EdgeInsets.all(8.0)".
  final EdgeInsetsGeometry? padding;

  /// Assign a [rotateType] to set the rotation type of the [child] widget.
  ///
  /// Possible rotation types are:
  ///
  /// [RotateType.full] => To rotate the [child] widget completely.
  ///
  /// [RotateType.semi] => To rotate the [child] widget 180 degrees
  /// and rotate back.
  ///
  /// [RotateType.quarter] => To rotate the [child] widget 90 degrees
  /// and rotate back.
  final RotateType rotateType;

  /// A variable to determine the direction of the animation.
  ///
  /// The default value is "true".
  final bool clockwise;

  /// A variable to determine the [duration] of the animation.
  ///
  /// The default value is "const Duration(milliseconds: 300)".
  final Duration duration;

  /// A variable to set the [curve] of the animation.
  ///
  /// The default value is "Curves.easeInOut".
  ///
  /// For more info please check the official documentation related
  /// to this topic:
  ///
  /// https://api.flutter.dev/flutter/animation/Curves-class.html
  final Curve curve;

  /// Assign a [shape] to set the shape of the button.
  ///
  /// The default value is "ButtonShape.rectangle".
  ///
  /// Possible shapes are:
  ///
  /// [ButtonShape.rectangle] => To give this button a rectangle or
  /// square shape.
  ///
  /// [ButtonShape.circle] => To give this button a circle shape.
  final ButtonShape shape;

  /// A variable to set the [elevation] of the button.
  ///
  /// The default value is "0.0".
  final double elevation;

  /// A variable to set the [shadowColor] of the button.
  ///
  /// Please note that this color will be visible only if the
  /// [elevation]'s value greater than 0.
  final Color? shadowColor;

  /// A variable to set the [background] color of the button.
  ///
  /// If you want to show only the child widget that you provide, you
  /// should set this [background] color to "Colors.transparent".
  final Color? background;

  /// A variable to set the [borderRadius] of the button.
  ///
  /// The default value is "0".
  ///
  /// Please note that this [borderRadius] will be applied to the button
  /// only if its shape is "ButtonShape.rectangle".
  final double? borderRadius;

  @override
  State<RotatingIconButton> createState() => _RotatingIconButtonState();
}

class _RotatingIconButtonState extends State<RotatingIconButton>
    with SingleTickerProviderStateMixin {
  late final AnimationController controller;
  late final Animation<double> animation;
  late double elevation;

  @override
  void initState() {
    super.initState();
    initController();
    addListener();
    initAnimation();
    setElevation();
  }

  void initController() {
    controller = AnimationController(
      vsync: this,
      duration: widget.duration,
      reverseDuration: widget.duration,
    );
  }

  void addListener() => controller.addStatusListener(listener);

  void listener(AnimationStatus status) {
    if (!mounted) return;
    if (widget.rotateType == RotateType.full) return;
    if (status != AnimationStatus.completed) return;

    controller.reverse();
  }

  void initAnimation() {
    late final double angle;

    switch (widget.rotateType) {
      case RotateType.quarter:
        angle = pi / 2;
        break;
      case RotateType.semi:
        angle = pi;
        break;
      case RotateType.full:
        angle = 2 * pi;
        break;
    }

    animation = Tween<double>(begin: 0, end: widget.clockwise ? angle : -angle)
        .animate(CurvedAnimation(parent: controller, curve: widget.curve));
  }

  void setElevation() => elevation = widget.elevation;

  @override
  void dispose() {
    controller
      ..removeStatusListener(listener)
      ..dispose();
    super.dispose();
  }

  void onTap() {
    if (!mounted) return;

    widget.onTap();
    controller.forward(from: 0.0);
  }

  void onTapDown(TapDownDetails tapDownDetails) {
    if (!mounted) return;
    if (widget.elevation == 0) return;

    setState(() => elevation = 0);
  }

  void onTapUp(TapUpDetails tapUpDetails) {
    if (!mounted) return;
    if (widget.elevation == 0) return;

    setState(setElevation);
  }

  void onTapCancel() {
    if (!mounted) return;
    if (widget.elevation == 0) return;

    setState(setElevation);
  }

  @override
  Widget build(BuildContext context) => GestureDetector(
        onTap: onTap,
        onTapDown: onTapDown,
        onTapUp: onTapUp,
        onTapCancel: onTapCancel,
        child: Material(
          color: widget.background,
          shape: getShape(),
          elevation: elevation,
          shadowColor: widget.shadowColor,
          child: Padding(
            padding: widget.padding ?? const EdgeInsets.all(8.0),
            child: AnimatedBuilder(
              animation: animation,
              builder: (context, child) =>
                  Transform.rotate(angle: animation.value, child: widget.child),
            ),
          ),
        ),
      );

  ShapeBorder getShape() {
    switch (widget.shape) {
      case ButtonShape.rectangle:
        return RoundedRectangleBorder(
          borderRadius: widget.borderRadius != null
              ? BorderRadius.circular(widget.borderRadius!)
              : BorderRadius.zero,
        );
      case ButtonShape.circle:
        return const CircleBorder();
    }
  }
}


example/mixed_buttons/linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "mixed_buttons");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "mixed_buttons");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


example/mixed_buttons/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"mixed_buttons", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/mixed_buttons/lib/main.dart
// Copyright 2022 The RotatingIconButton Author. All rights reserved.
// Use of this source code is governed by a MIT-style license that
// can be found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:rotating_icon_button/rotating_icon_button.dart';

void main() {
  runApp(const ExampleApp());
}

class ExampleApp extends StatelessWidget {
  const ExampleApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) => MaterialApp(
        title: 'Mixed Buttons',
        debugShowCheckedModeBanner: false,
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: const MixedButtons(),
      );
}

class MixedButtons extends StatefulWidget {
  const MixedButtons({Key? key}) : super(key: key);

  @override
  State<MixedButtons> createState() => _MixedButtonsState();
}

class _MixedButtonsState extends State<MixedButtons> {
  @override
  Widget build(BuildContext context) => Scaffold(
        backgroundColor: Colors.cyanAccent,
        appBar: AppBar(
          title: const Text(
            'Mixed Buttons',
            style: TextStyle(color: Colors.black),
          ),
          backgroundColor: Colors.cyanAccent,
        ),
        body: SafeArea(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.spaceEvenly,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              buildFirstRow(),
              buildSecondRow(),
              buildThirdRow(),
              buildFourthRow(),
              buildFifthdRow(),
              buildSixthRow(),
            ],
          ),
        ),
      );

  Widget buildFirstRow() => Row(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          RotatingIconButton(
            onTap: () {},
            background: Colors.transparent,
            clockwise: false,
            child: const Icon(
              Icons.settings,
              color: Colors.black,
              size: 30,
            ),
          ),
          RotatingIconButton(
            onTap: () {},
            background: Colors.transparent,
            padding: const EdgeInsets.all(2.0),
            shape: ButtonShape.circle,
            child: const Icon(
              Icons.settings_outlined,
              color: Colors.red,
              size: 30,
            ),
          ),
        ],
      );

  Row buildSecondRow() => Row(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          RotatingIconButton(
            onTap: () {},
            elevation: 10.0,
            background: Colors.amber,
            clockwise: false,
            child: const Icon(
              Icons.search,
              color: Colors.indigo,
            ),
          ),
          RotatingIconButton(
            onTap: () {},
            elevation: 10.0,
            background: Colors.indigo,
            shape: ButtonShape.circle,
            child: const Icon(
              Icons.search,
              color: Colors.amber,
            ),
          ),
        ],
      );

  Row buildThirdRow() => Row(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          RotatingIconButton(
            onTap: () {},
            elevation: 20.0,
            padding: const EdgeInsets.symmetric(
              vertical: 10,
              horizontal: 40,
            ),
            rotateType: RotateType.semi,
            background: Colors.deepPurple,
            clockwise: false,
            child: const Icon(
              Icons.attach_file,
              color: Colors.amberAccent,
            ),
          ),
          RotatingIconButton(
            onTap: () {},
            elevation: 20.0,
            background: Colors.amberAccent,
            shape: ButtonShape.circle,
            rotateType: RotateType.semi,
            padding: const EdgeInsets.symmetric(
              vertical: 20,
              horizontal: 20,
            ),
            child: const Icon(
              Icons.attach_file,
              color: Colors.deepPurple,
            ),
          ),
        ],
      );

  Row buildFourthRow() => Row(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          RotatingIconButton(
            onTap: () {},
            elevation: 10.0,
            shadowColor: Colors.pink,
            borderRadius: 20.0,
            rotateType: RotateType.semi,
            clockwise: false,
            padding: const EdgeInsets.symmetric(
              vertical: 10,
              horizontal: 40,
            ),
            background: Colors.grey,
            child: const Icon(
              Icons.navigation,
              color: Colors.pink,
            ),
          ),
          RotatingIconButton(
            onTap: () {},
            elevation: 10.0,
            shadowColor: Colors.pink,
            background: Colors.pink,
            rotateType: RotateType.semi,
            shape: ButtonShape.circle,
            padding: const EdgeInsets.symmetric(
              vertical: 20,
              horizontal: 20,
            ),
            child: const Icon(
              Icons.navigation,
              color: Colors.grey,
            ),
          ),
        ],
      );

  Row buildFifthdRow() => Row(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          RotatingIconButton(
            onTap: () {},
            elevation: 10.0,
            borderRadius: 10.0,
            rotateType: RotateType.full,
            duration: const Duration(milliseconds: 600),
            background: Colors.amber,
            clockwise: false,
            child: const Icon(
              Icons.rotate_left,
              color: Colors.indigo,
            ),
          ),
          RotatingIconButton(
            onTap: () {},
            elevation: 10.0,
            rotateType: RotateType.full,
            duration: const Duration(milliseconds: 600),
            background: Colors.indigo,
            shape: ButtonShape.circle,
            child: const Icon(
              Icons.rotate_right,
              color: Colors.amber,
            ),
          ),
        ],
      );

  Widget buildSixthRow() => Row(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          Container(
            color: Colors.brown,
            padding: const EdgeInsets.all(2.0),
            child: RotatingIconButton(
              onTap: () {},
              background: Colors.cyanAccent,
              padding: EdgeInsets.zero,
              clockwise: false,
              child: const Icon(
                Icons.add,
                color: Colors.brown,
              ),
            ),
          ),
          RotatingIconButton(
            onTap: () {},
            background: Colors.deepOrange,
            padding: const EdgeInsets.all(2.0),
            shape: ButtonShape.circle,
            child: Container(
              decoration: const BoxDecoration(
                shape: BoxShape.circle,
                color: Colors.cyanAccent,
              ),
              child: const Icon(
                Icons.add,
                color: Colors.deepOrange,
              ),
            ),
          ),
        ],
      );
}
